akd/src/storage/memory.rs

```diff
@@ -100,12 +104,20 @@ impl Storage for AsyncInMemoryDatabase {
         }
 
         if let DbRecord::ValueState(value_state) = &record {
-            let mut guard = self.user_info.write().await;
-            let username = value_state.username.0.clone();
-            guard
-                .entry(username)
-                .or_insert_with(Vec::new)
-                .push(value_state.clone());
+            let mut u_guard = self.user_info.write().await;
+            let username = value_state.username.to_vec();
+            match u_guard.get(&username) {
+                Some(old_states) => {
+                    let mut new_states = old_states.clone();
+                    new_states.insert(value_state.epoch, value_state.clone());
+                    u_guard.insert(username, new_states);
+                }
+                None => {
+                    let mut new_map = HashMap::new();
+                    new_map.insert(value_state.epoch, value_state.clone());
+                    u_guard.insert(username, new_map);
+                }
+            }
```

Comment:
Maybe you've already tried this, but does `entry()` and `or_insert()` work here?